components/encoder/as5048a.go

```
@@ -0,0 +1,296 @@
+	if len(connType) == 0 {
+		// default to I2C until SPI support is implemented
+		conf.ConnectionType = i2cConn
+		// return nil, errors.New("must specify connection type")
+	}
+	_, isSupported := supportedConnections[connType]
+	if !isSupported {
+		return nil, errors.Errorf("%s is not a supported connection type", connType)
+	}
+	if connType == i2cConn {
+		if len(conf.BoardName) == 0 {
+			return nil, errors.New("expected nonempty board")
+		}
+		err := conf.i2cAttrConfig.ValidateI2C(path)
+		if err != nil {
+			return nil, err
+		}
+		deps = append(deps, conf.BoardName)
+	}
+
+	return deps, nil
+}
+
+type i2cAttrConfig struct {
+	I2CBus  string `json:"i2c_bus"`
+	I2CAddr int    `json:"i2c_addr"`
+}
+
+// ValidateI2C ensures all parts of the config are valid.
+func (cfg *i2cAttrConfig) ValidateI2C(path string) error {
+	if cfg.I2CBus == "" {
+		return utils.NewConfigValidationFieldRequiredError(path, "i2c_bus")
+	}
+	if cfg.I2CAddr == 0 {
+		return utils.NewConfigValidationFieldRequiredError(path, "i2c_addr")
+	}
+
+	return nil
+}
+
+// AS5048A is a struct representing an instance of an AS5048A
+// Hall-effect encoder.
+type AS5048A struct {
+	mu                      sync.RWMutex
+	logger                  golog.Logger
+	position                float64
+	positionOffset          float64
+	rotations               int
+	connectionType          string
+	i2cHandle               board.I2CHandle
+	cancelCtx               context.Context
+	cancel                  context.CancelFunc
+	activeBackgroundWorkers sync.WaitGroup
+	generic.Unimplemented
+}
+
+func (enc *AS5048A) startPositionLoop(ctx context.Context) error {
+	if err := enc.Reset(ctx, 0.0, map[string]interface{}{}); err != nil {
+		return err
+	}
+	enc.activeBackgroundWorkers.Add(1)
+	utils.ManagedGo(func() {
+		for {
+			if enc.cancelCtx.Err() != nil {
+				return
+			}
+			if err := enc.updatePosition(ctx); err != nil {
+				enc.logger.Errorf("error in position loop: %s", err.Error())
```

Why not just return the error?

---

for one, you can't return from inside the anonymous function. But also, isn't it too fragile for the entire loop to close with a single failed reading?

---

You're right about the loop logic, my bad. Inform the user that we're continuing to read after error.